lxd/migration/wsproto.go

```diff
@@ -0,0 +1,71 @@
+package migration
+
+import (
+	"fmt"
+	"io/ioutil"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/gorilla/websocket"
+
+	"github.com/lxc/lxd/shared"
+)
+
+// ProtoRecv gets a protobuf message from a websocket
+func ProtoRecv(ws *websocket.Conn, msg proto.Message) error {
+	mt, r, err := ws.NextReader()
+	if err != nil {
+		return err
+	}
+
+	if mt != websocket.BinaryMessage {
+		return fmt.Errorf("Only binary messages allowed")
+	}
+
+	buf, err := ioutil.ReadAll(r)
+	if err != nil {
+		return err
+	}
+
+	err = proto.Unmarshal(buf, msg)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ProtoSend sends a protobuf message over a websocket
+func ProtoSend(ws *websocket.Conn, msg proto.Message) error {
+	w, err := ws.NextWriter(websocket.BinaryMessage)
+	if err != nil {
+		return err
+	}
+	defer w.Close()
+
+	data, err := proto.Marshal(msg)
+	if err != nil {
+		return err
+	}
+
+	err = shared.WriteAll(w, data)
```

Comment:
Would we benefit from using our EAGAIN writer here as well?

Reply:
I don't think so, "w" over there is a gorilla websocket file object which I don't believe can even return EAGAIN (as it's not a fd).